File: code_gasoline_france/analysis/analysis_dispersion/pairs/rr_vs_distance_regressions.py
﻿#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import print_function
import add_to_path
from add_to_path import path_data
from generic_master_price import *
from generic_master_info import *
from generic_competition import *
import time

path_dir_built = os.path.join(path_data,
                              u'data_gasoline',
                              u'data_built',
                              u'data_scraped_2011_2014')
path_dir_built_csv = os.path.join(path_dir_built, u'data_csv')
path_dir_built_json = os.path.join(path_dir_built, u'data_json')
path_dir_built_graphs = os.path.join(path_dir_built, u'data_graphs')

path_dir_built_dis = os.path.join(path_data,
                                  u'data_gasoline',
                                  u'data_built',
                                  u'data_dispersion')
path_dir_built_dis_csv = os.path.join(path_dir_built_dis, u'data_csv')
path_dir_built_dis_json = os.path.join(path_dir_built_dis, u'data_json')

pd.set_option('float_format', '{:,.3f}'.format)
format_float_int = lambda x: '{:10,.0f}'.format(x)
format_float_float = lambda x: '{:10,.3f}'.format(x)

# ################
# LOAD DATA
# ################

# DF INFO
df_info = pd.read_csv(os.path.join(path_dir_built_csv,
                                   'df_station_info_final.csv'),
                      encoding = 'utf-8',
                      dtype = {'id_station' : str,
                               'adr_zip' : str,
                               'adr_dpt' : str,
                               'ci_1' : str,
                               'ci_ardt_1' :str,
                               'ci_2' : str,
                               'ci_ardt_2' : str,
                               'dpt' : str})
df_info.set_index('id_station', inplace = True)

# DF STATION STATS
df_station_stats = pd.read_csv(os.path.join(path_dir_built_csv,
                                            'df_station_stats.csv'),
                               encoding = 'utf-8',
                               dtype = {'id_station' : str})
df_station_stats.set_index('id_station', inplace = True)

# CLOSE STATIONS
dict_ls_close = dec_json(os.path.join(path_dir_built_json,
                                      'dict_ls_close.json'))

# DF PRICES
df_prices_ttc = pd.read_csv(os.path.join(path_dir_built_csv,
                                         'df_prices_ttc_final.csv'),
                        parse_dates = ['date'])
df_prices_ttc.set_index('date', inplace = True)

df_prices_ht = pd.read_csv(os.path.join(path_dir_built_csv,
                                        'df_prices_ht_final.csv'),
                        parse_dates = ['date'])
df_prices_ht.set_index('date', inplace = True)

df_prices_cl = pd.read_csv(os.path.join(path_dir_built_csv,
                                        'df_cleaned_prices.csv'),
                          parse_dates = ['date'])
df_prices_cl.set_index('date', inplace = True)

# basic station filter (geo scope or insufficient obs.)
df_filter = df_station_stats[~((df_station_stats['pct_chge'] < 0.03) |\
                               (df_station_stats['nb_valid'] < 90))]
ls_keep_ids = list(set(df_filter.index).intersection(\
                     set(df_info[(df_info['highway'] != 1) &\
                                 (df_info['reg'] != 'corse')].index)))
ls_drop_ids = list(set(df_prices_ttc.columns).difference(set(ls_keep_ids)))
df_prices_ttc[ls_drop_ids] = np.nan
df_prices_ht[ls_drop_ids] = np.nan
# highway stations may not be in df_prices_cl (no pbm here)
ls_drop_ids_cl = list(set(df_prices_cl.columns).difference(set(ls_keep_ids)))
df_prices_cl[ls_drop_ids_cl] = np.nan
#df_info = df_info.ix[ls_keep_ids]
#df_station_stats = df_station_stats.ix[ls_keep_ids]

# DF PAIRS
ls_dtype_temp = ['id', 'ci_ardt_1']
dict_dtype = dict([('%s_1' %x, str) for x in ls_dtype_temp] +\
                  [('%s_2' %x, str) for x in ls_dtype_temp])
df_pairs = pd.read_csv(os.path.join(path_dir_built_dis_csv,
                                    'df_pair_final.csv'),
              encoding = 'utf-8',
              dtype = dict_dtype)

# basic pair filter (insufficient obs.)
df_pairs = df_pairs[~((df_pairs['nb_spread'] < 90) &\
                      (df_pairs['nb_ctd_both'] < 90))]

# todo? harmonize pct i.e. * 100

# LISTS FOR DISPLAY
lsd = ['id_1', 'id_2', 'distance', 'group_last_1', 'group_last_2']
lsd_rr = ['rr_1', 'rr_2', 'rr_3', 'rr_4', 'rr_5', '5<rr<=20', 'rr>20']

# #############
# PREPARE DATA
# #############

# RESTRICT CATEGORY: PRICES AND MARGIN CHGE
df_pairs_all = df_pairs.copy()
price_cat = 'no_mc' # 'residuals_no_mc'
print(u'Prices used : {:s}'.format(price_cat))
df_pairs = df_pairs[df_pairs['cat'] == price_cat].copy()

# robustness check with idf: 1 km max
ls_dense_dpts = [75, 92, 93, 94]
df_pairs = df_pairs[~((((df_pairs['dpt_1'].isin(ls_dense_dpts)) |\
                        (df_pairs['dpt_2'].isin(ls_dense_dpts))) &\
                       (df_pairs['distance'] > 1)))]

## robustness check keep closest competitor
#df_pairs.sort(['id_1', 'distance'], ascending = True, inplace = True)
#df_pairs.drop_duplicates('id_1', inplace = True, take_last = False)
# could also collect closest for each id_2 and filter further
# - id_1 can have closer competitor as an id_2
# - duplicates in id_2 (that can be solved also but drops too much)
#df_pairs.sort(['id_2', 'distance'], ascending = True, inplace = True)
#df_pairs.drop_duplicates('id_2', inplace = True, take_last = False)
## - too many drops: end ids always listed as id_2 disappear... etc.

### robustness check: rr>20 == 0
#df_pairs = df_pairs[(df_pairs['rr>20'] == 0)]
#df_pairs = df_pairs[(df_pairs['mean_rr_len'] <= 21)]

# COMPETITORS VS. SAME GROUP
df_pair_same =\
  df_pairs[(df_pairs['group_1'] == df_pairs['group_2']) &\
           (df_pairs['group_last_1'] == df_pairs['group_last_2'])].copy()
df_pair_comp =\
  df_pairs[(df_pairs['group_1'] != df_pairs['group_2']) &\
           (df_pairs['group_last_1'] != df_pairs['group_last_2'])].copy()

# DIFFERENTIATED VS. NON DIFFERENTIATED
diff_bound = 0.01
df_pair_same_nd = df_pair_same[df_pair_same['abs_mean_spread'] <= diff_bound]
df_pair_same_d  = df_pair_same[df_pair_same['abs_mean_spread'] > diff_bound]
df_pair_comp_nd = df_pair_comp[df_pair_comp['abs_mean_spread'] <= diff_bound]
df_pair_comp_d  = df_pair_comp[df_pair_comp['abs_mean_spread'] > diff_bound]

# DICT OF DFS
# pairs without spread restriction
# (keep pairs with group change in any)
dict_pair_comp = {'any' : df_pair_comp}
for k in df_pair_comp['pair_type'].unique():
  dict_pair_comp[k] = df_pair_comp[df_pair_comp['pair_type'] == k]
# low spread pairs
dict_pair_comp_nd = {}
for df_temp_title, df_temp in dict_pair_comp.items():
  dict_pair_comp_nd[df_temp_title] =\
      df_temp[df_temp['abs_mean_spread'] <= diff_bound]

# CREATE SAME CORNER VARIABLES
df_pairs['sc_500'] = 0
df_pairs.loc[df_pairs['distance'] <= 0.5, 'sc_500'] = 1
df_pairs['sc_750'] = 0
df_pairs.loc[df_pairs['distance'] <= 0.75, 'sc_750'] = 1
df_pairs['sc_1000'] = 0
df_pairs.loc[df_pairs['distance'] <= 1, 'sc_1000'] = 1

# ##################
# REGRESSION
# ##################

# REGRESSION FORMULAS AND PARAMETERS
ls_dist_ols_formulas = ['abs_mean_spread ~ distance',
                        'mean_abs_spread ~ distance',
                        'pct_rr ~ distance',
                        'std_spread ~ distance']

dist_reg = 1000

ls_sc_ols_formulas = ['abs_mean_spread ~ sc_{:d}'.format(dist_reg),
                      'mean_abs_spread ~ sc_{:d}'.format(dist_reg),
                      'pct_rr ~ sc_{:d}'.format(dist_reg),
                      'std_spread ~ sc_{:d}'.format(dist_reg)]

# from statsmodels.regression.quantile_regression import QuantReg
ls_quantiles = [0.25, 0.5, 0.75, 0.90]

#mod = smf.quantreg('pct_rr~distance', df_ppd_reg[~pd.isnull(df_ppd_reg['pct_rr'])])
#res = mod.fit(q=.5)
#print res.summary()
## Following: need to add constant to make it equivalent
#res_alt = QuantReg(df_ppd_reg['pct_rr_nozero'], df_ppd_reg['sc_500']).fit(0.5)
# So far: need to add "resid[resid == 0] = .000001" in quantreg line 171-3 to have it run

# check => https://groups.google.com/forum/?hl=fr#!topic/pystatsmodels/XnXu_k1h-gc

# Output data to csv to run quantile regressions in R
df_pair_comp.to_csv(os.path.join(path_dir_built_dis_csv,
                                 'df_pair_comp_{:s}.csv'.format(price_cat)),
                    float_format= '%.4f',
                    encoding = 'utf-8',
                    index = False)

# REWRITE OUTPUT REGS

# FOR OLS AND QR
# - provide var names
# - provide stats to output for these vars
# - provide general reg stats to output
# - create dataframe with one row per reg: index is specification or part of it?

def format_ls_reg_fits_to_df(ls_tup_reg_fit, ls_var_names):
  ls_var_attrs = ['params', 'bse', 'tvalues', 'pvalues']
  ls_reg_attrs = [] #['nobs', 'rsquared', 'rsquared_adj']
  ls_tup_vars = [(var_name, var_attr) for var_attr in ls_var_attrs\
                      for var_name in ls_var_names]
  ls_index, ls_rows = [], []
  for reg_title, reg_fit in ls_tup_reg_fit:
    ls_var_stats = [getattr(reg_fit, var_attr).ix[var_name]\
                      for (var_name, var_attr) in ls_tup_vars]
    ls_reg_stats = [getattr(reg_fit, reg_attr) for reg_attr in ls_reg_attrs]
    ls_rows.append(ls_reg_stats + ls_var_stats)
    ls_index.append(reg_title)
  df_reg_fits = pd.DataFrame(ls_rows,
                             columns = ls_reg_attrs +\
                                       ['_'.join(x) for x in ls_tup_vars],
                             index = ls_index)
  return df_reg_fits

ls_loop_df_ppd_regs = [('All', dict_pair_comp['any']),
                       ('Sup', dict_pair_comp['sup']),
                       ('Nd All', dict_pair_comp_nd['any']),
                       ('Nd Oil&Ind', dict_pair_comp_nd['oil&ind']),
                       ('Nd Sup', dict_pair_comp_nd['sup']),
                       ('Sup Dis', dict_pair_comp['sup_dis']),
                       ('Nd Sup Dis', dict_pair_comp_nd['sup_dis']),
                       ('Oil&Ind', dict_pair_comp['oil&ind'])]

# Some issues with current implementation of quantreg => R
# Work with dist 1000 until Oil Ind...

for df_ppd_title, df_ppd_reg in ls_loop_df_ppd_regs:
  df_ppd_reg = df_ppd_reg[(df_ppd_reg['distance'] <= 3)].copy()
  #df_ppd_reg.loc[df_ppd_reg['pct_rr'] == 0, 'pct_rr'] = 0.0001
  #df_ppd_reg.loc[df_ppd_reg['distance'] == 0, 'distance'] = 0.0001
  print()
  print(df_ppd_title)
  # Simple ols
  ls_dis_ols_fits = [(str_formula,
                     smf.ols(formula = str_formula, data = df_ppd_reg).fit())\
                       for str_formula in ls_sc_ols_formulas]
  df_dis_ols_res = format_ls_reg_fits_to_df(ls_dis_ols_fits, ['sc_{:d}'.format(dist_reg)])
  print()
  print(df_dis_ols_res.to_string())
  # QRs rank reversals
  ls_rr_qr_fits  = [('rr_sc_Q{:.2f}'.format(quantile),
                     smf.quantreg('pct_rr~sc_{:d}'.format(dist_reg),
                                  data = df_ppd_reg).fit(quantile))\
                       for quantile in ls_quantiles]
  df_rr_qr_fits = format_ls_reg_fits_to_df(ls_rr_qr_fits, ['sc_{:d}'.format(dist_reg)])
  print()
  print(df_rr_qr_fits.to_string())
  # QRs standard deviation
  ls_std_qr_fits  = [('std_sc_Q{:.2f}'.format(quantile),
                       smf.quantreg('std_spread~sc_{:d}'.format(dist_reg),
                                    data = df_ppd_reg).fit(quantile))\
                         for quantile in ls_quantiles]
  df_std_qr_fits = format_ls_reg_fits_to_df(ls_std_qr_fits, ['sc_{:d}'.format(dist_reg)])
  print()
  print(df_std_qr_fits.to_string())

File: code_gasoline_france/analysis/analysis_dispersion/pairs/rr_vs_distance_regressions.R
library(quantreg)
options("scipen"=30, "digits"=4)
digits.option <- 4

path.data <- 'C:/Users/etna/Desktop/Etienne_work/Data'
if (file.exists(file.path(path.data))){
    path.data <- path.data
}else{
    path.data <- '//ulysse/users/echamayou/Bureau/Etienne_work/Data'
}
path.data.disp <- file.path(path.data, 'data_gasoline/data_built/data_dispersion/data_csv')
path.file <- file.path(path.data.disp, 'df_pair_comp_no_mc.csv')

market.dist <- 5

sink(file.path(path.data.disp,
               paste(c('R_pair_regs_', market.dist, 'km.txt'), collapse = '')))

data <- read.csv(path.file)
ppd <- data
#ppd <- subset(data, !is.na(pct_rr) &
#                    !is.na(abs_mean_spread) &
#                    !is.na(distance))
#ppd <- na.omit(data)

## Prepare data
#ppd <- subset(ppd, cat == 'no_mc')
#ppd <- subset(ppd, (group_1 != group_2) & (group_last_1 != group_last_2))
#ppd$sc_500  <- ifelse(ppd$distance <= 0.50, 1, 0)
#ppd$sc_750  <- ifelse(ppd$distance <= 0.75, 1, 0)
#ppd$sc_1000 <- ifelse(ppd$distance <= 1.00, 1, 0)

ppd <- subset(ppd, distance <= market.dist)
ppd.nodiff <- subset(ppd, abs_mean_spread <= 0.01)

# Preliminary OLS

oilind <- c('OIL', 'INDEPENDENT')
ls.all <- list(ppd = ppd,
               ppd.sup = subset(ppd, (group_type_1 == 'SUP') &
                                     (group_type_2 == 'SUP')),
               ppd.oil.ind = subset(ppd, (is.element(group_type_1, oilind)) &
                                         (is.element(group_type_2, oilind))),
               ppd.dis = subset(ppd, (group_type_1 == 'DIS') &
                                     (group_type_2 == 'DIS')),
               ppd.sup.dis = subset(ppd, ((group_type_1 == 'SUP') &
                                          (group_type_2 == 'DIS')) |
                                         ((group_type_1 == 'DIS') &
                                          (group_type_2 == 'SUP'))),
               ppd.oil.sup = subset(ppd, ((group_type_1 == 'SUP') &
                                          (is.element(group_type_2, oilind))) |
                                         ((is.element(group_type_1, oilind)) &
                                          (group_type_2 == 'SUP'))))

ls.nd <- lapply(ls.all, function(ppdtemp){
  ppdtemp <- subset(ppdtemp, abs_mean_spread <= 0.01)
})

ls.temp <- ls.all

print('Preliminary OLS regs')
ls.temp.names <- names(ls.temp)
for (name in ls.temp.names) {
  cat(replicate(20, '-'))
  cat("\n", name, "\n")
  ppdtemp <- ls.temp[[name]]
  fitmeanspread <- lm(abs_mean_spread ~ distance, ppdtemp)
  fitpctrrdist <- lm(pct_rr ~ distance, ppdtemp)
  cat("\n", "abs mean spread", "\n")
  #print(summary(fitmeanspread, digits = digits.option))
  print(summary(fitmeanspread, digits = digits.option)$coefficients[2,])
  cat("\n", "pct rr", "\n")
  #print(summary(fitpctrrdist, digits = digits.option))
  print(summary(fitpctrrdist, digits = digits.option)$coefficients[2,])
  
  cat("\n", "OLS and Quantile regs (loop on same corner dist and quantiles)", "\n")
  sc_distance <- 1000
  sc <- paste("sc_", sc_distance, sep="")
  res_lm <- lm(pct_rr ~ get(sc), ppdtemp)
  #print(summary(res_lm))
  #print('Simple OLS')
  #print(summary(res_lm, digits = digits.option)$coefficients[2,1:4])
  ls.qres <- list(summary(res_lm, digits = digits.option)$coefficients[2,1:4])
  for(tau_option in c(0.5, 0.75)) {
    res_rq <- rq(formula = pct_rr ~ get(sc), tau = tau_option, data = ppdtemp)
    #cat('\n')
    #print(summary(res_rq))
    #print(paste('Quantile:', tau_option))
    #print(summary(res_rq, digits = digits.option)$coefficients[2,1:4])
    ls.qres[[length(ls.qres)+1]] <- summary(res_rq)$coefficients[2,1:4]
  }
  df.qres <- data.frame(x = matrix(unlist(ls.qres), nrow=length(ls.qres), byrow=T))
  colnames(df.qres) <- c('coef', 'std', 'tval', 'pval')
  #rownames(df.qres) <- c('OLS', 'Q25', 'Q50', 'Q75', 'Q90')
  rownames(df.qres) <- c('OLS', 'Q50', 'Q75')
  print(df.qres)
  
}

# OLS and quantile regressions with same corner variables



## Kolmogorov Smirnov tests
## https://stat.ethz.ch/R-manual/R-devel/library/stats/html/ks.test.html
#
#cat('\n')
#print('KS test (loop on same corner dist and hypotheses)')
#for(sc_distance in c(0.5, 0.75, 1.00)) {
#  cat('\n')
#  print(paste(replicate(40, '-'), collapse = ''))
#  print(paste('Distance:', sc_distance))
#  for(alternative_option in c('two.sided', 'less', 'greater'))
#  {
#    rr.close <- ppd.nodiff[which(ppd.nodiff$distance <= sc_distance), c('pct_rr')]
#    rr.further <- ppd.nodiff[which(ppd.nodiff$distance > sc_distance & ppd.nodiff$distance <= market.dist), c('pct_rr')]
#    res_ks <- ks.test(rr.close, rr.further, alternative = c(alternative_option))
#    print(res_ks)
#  }
#}
#

sink()
#unlink(file.path(path.data.disp, 'R_pair_regs.txt'))
closeAllConnections()

## BACKUP
#cat("\n", "OLS and Quantile regs (loop on same corner dist and quantiles)", "\n")
#for(sc_distance in c(0.5, 0.75, 1.00)) {
#  cat('\n')
#  print(paste(replicate(40, '-'), collapse = ''))
#  print(paste('Distance:', sc_distance))
#  sc <- paste("sc_", 750, sep="")
#  res_lm <- lm(pct_rr ~ get(sc), ppd.nodiff)
#  #print(summary(res_lm))
#  print('Simple OLS')
#  print(summary(res_lm, digits = digits.option)$coefficients[2,1:4])
#  for(tau_option in c(0.25, 0.5, 0.75, 0.90))
#  {
#    res_rq <- rq(formula = pct_rr ~ get(sc), tau = tau_option, data = ppd.nodiff)
#    cat('\n')
#    # print(summary(res_rq))
#    print(paste('Quantile:', tau_option))
#    print(summary(res_rq, digits = digits.option)$coefficients[2,1:4])
#  }
#}
